---
title: "Human CRPS Combined Cohorts 16S Analysis of RDP and SILVA Annotated Data"
author: "Rachel Rodgers"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      warning = FALSE,
                      message = FALSE)
```

```{r load-libraries, message=FALSE, include=FALSE}
library("shiny")
library("phyloseq")
library("vegan")
library("gridExtra")
library("knitr")
library("plotly")
library("microbiome")
library("data.table")
library("ggrepel")
library("ggpubr")
library("scales")
library("plyr")
library("DT")
library("picante")
library("gridExtra")
library("DESeq2")
library("tidyverse")

source("./shared_R_scripts/Helper_Functions.R")
source("./shared_R_scripts/BoxPlotHelpers.R")
source("./shared_R_scripts/BiomarkerHelpers.R")

options(shiny.sanitize.errors = FALSE)
```

## Overview {.tabset}

This is a 16S analysis of human CRPS data from across five sequencing runs:

  1. 190201_Baldridge_CGMA
  2. 190318_Baldridge_CGMA
  3. 190523_Baldridge_CGMA_V4
  4. 200127_Baldridge_CGMA_16S
  5. 201019_LaraC_V4_Mouse_Human_CRPS
  
The taxa have been annotated with both RDP (v16) and SILVA (v138.1) training sets.
The analyses of both data sets will be compared.

<br>
<br>

### **QC**

#### **Taxa Filtering**

<br>

Non-bacterial taxa need to be removed from the data set prior to further analysis.
The RDP data set contains taxa from the Phylum Cyanobacteria/Chloroplast, while 
the SILVA data contains taxa from the Kingdom Archaea and the Phylum Cyanobacteria.
Those taxa will be removed here.

```{r filter-non-bacterial, results='hide'}
# RDP
physeqMergedRDP <- readRDS("../data/RDP/RDataObjects/physeqObjects/physeqMergedRDP.RDS")
physeqMergedRDP # 2115 x 67

physeqBacteriaRDP <- physeqMergedRDP %>% 
  subset_taxa(Kingdom == "Bacteria" & Phylum != "Cyanobacteria/Chloroplast")
physeqBacteriaRDP # 2061 x 67

# SILVA
physeqMergedSILVA <- readRDS("../data/SILVA/RDataObjects/physeqObjects/physeqMergedSILVA.RDS")
physeqMergedSILVA # 2668 x 67

physeqBacteriaSILVA <- physeqMergedSILVA %>% 
  subset_taxa(Kingdom == "Bacteria" & Phylum != "Cyanobacteria")
physeqBacteriaSILVA # 2640 x 67
```

Removal of non-bacterial taxa from the RDP data set reduces the total number of 
taxa from 2,115 to 2,061. In the SILVA data, total taxa are reduced from 2,668
to 2,640.

```{r physeqList}
# store the different physeq objects in a list
physeqList <- list("rdp" = physeqBacteriaRDP,
                   "silva" = physeqBacteriaSILVA)

sampleDataList <- map(.x = physeqList,
                      .f = ~ as(sample_data(.), "data.frame"))

varsOfInterest <- c("Sample Type" = "Sample_Type",
                    "Household Type" = "bad_household",
                    "CRPS Type" = "Description")
```

<br>

---

#### **Read Count Distribution**

<br>

```{r reads-per-sample-summary}
numReadsSummary <- list("rdp" = summary(sampleDataList$rdp$readsPerSampleRDP),
                        "silva" = summary(sampleDataList$silva$readsPerSampleSILVA))
numReadsSummary
```

<br>

---

**By Library**

<br>

```{r read-counts-by-library}
sampleDataList <- map(.x = sampleDataList,
                      .f = ~ .x %>% 
                        mutate(across("run", as.character)))

readsByLib <- pmap(.l = list(df = sampleDataList,
                             yVar = c("rdp" = "readsPerSampleRDP",
                                      "silva" = "readsPerSampleSILVA"),
                             plotTitle = paste("Avg. Reads by Sequencing Run\n",
                                               toupper(names(sampleDataList)), 
                                               "Data")),
                   .f = PrettyBox,
                   xVar = "run",
                   statMethod = "kruskal.test",
                   colorVals = cbPaletteGrey,
                   label_x = "Sequencing Run",
                   label_y = "Reads Per Sample")

# Add horizontal line at 1k reads and label the low read samples on the plot
lowReadData <- map2(.x = sampleDataList,
                    .y = c("rdp" = "readsPerSampleRDP",
                           "silva" = "readsPerSampleSILVA"),
                    .f = ~ .x %>% 
                      filter(!!sym(.y) <= 1000))


readsByLibModified <- map2(.x = readsByLib,
                           .y = lowReadData,
                           .f = ~ .x +
                             geom_hline(yintercept = 1000, lty = "dashed") +
                             geom_text_repel(aes_string(label = "sample"),
                                             data = .y))

readsByLibModified

lowReadSamples <- map(.x = lowReadData, .f = ~ .x$sample)
```

There is a significant difference between the average reads per sample by library. 
The data will be rarefied to compensate for this.

In addition there are four BioBank Control samples with low (< 1000) reads 
which will be removed from the analysis:

  * Baldridge_D
  * Baldridge_R
  * Baldridge_U
  * Baldridge_X
  
```{r trim-physeq, results='hide'}
RemoveLowReadSamples <- function(df, physeq, numReadsCol, sampleNameCol,
                                 lowReadThresh = 1000) {
  # get the names of the samples that exceed threshold
  keepSamples <- df %>% 
    filter(!!sym(numReadsCol) > lowReadThresh) %>% 
    pull(!!sym(sampleNameCol)) %>% 
    as.character()
  
  # prune out the unwanted samples
  physeqTrimmed <- physeq %>% 
    prune_samples(samples = keepSamples) %>% 
    RemoveMissingTaxa()
  
}

numReadsColList <- c("rdp" = "readsPerSampleRDP", "silva" = "readsPerSampleSILVA")
sampleNameColList <- c("rdp" = "sample", "silva" = "sequencing_file_name")

physeqListTrimmed <- pmap(.l = list(physeq = physeqList,
                                    numReadsCol = numReadsColList,
                                    sampleNameCol = sampleNameColList,
                                    df = sampleDataList),
                          .f = RemoveLowReadSamples)
physeqListTrimmed

# extract sample data from the trimmed physeq objects
sampleDataTrimmed <- map(.x = physeqListTrimmed,
                         .f = ~ as(sample_data(.x), "data.frame"))
```

Removing these low read samples has decreased the total number of taxa in the 
RDP data from 2,061 to 2,047, and in the SILVA data from 2,640 to 2,096.

<br>

---

#### **Rarefy**

<br>

```{r physeqListRare, results='hide'}
physeqListRare <- map(.x = physeqListTrimmed,
                      .f = ~ rarefy_even_depth(.x, 
                                               rngseed = 3973127))
physeqListRare

sampleDataRare <- map(.x = physeqListRare,
                      .f = ~ as(sample_data(.x), "data.frame"))
```

Rarefying the data has reduced the total number of taxa in the RDP data set from
2,047 to 1,590 and in the SILVA data set from 2,096 to 1,600.

---

<br>

#### **Phyla Prevalence**

Checking the prevalence and abundance of different phyla to help with analyzing
biomarkers later in the workflow.

```{r phylaPrev}
phylaPrevRes <- map(.x = physeqListRare,
                    .f = ~ TaxRankPrevalence(.x, "Phylum"))
phylaPrevDFList <- map(.x = phylaPrevRes,
                       .f = ~ .x$prevalence_df)

phylaPrevPlots <- pmap(.l = list(prevDF = phylaPrevDFList,
                                 physeqObj = physeqListRare,
                                 myTitle = list("rdp" = "RDP",
                                                "silva" = "SILVA"),
                                 intercept_y = list("rdp" = 1/nsamples(physeqListRare$rdp),
                                                    "silva" = 1/nsamples(physeqListRare$silva))),
                       .f = PlotPhylaPrevalence)

phylaPrevPlots
```

<br>

### **Community Composition**

```{r reactive-community-composition}
#----- UI -----#

fluidRow(
  column(5,
         # select comparison
         selectInput(inputId = "communityComparison", 
                     label = "Select Comparison:",
                     choices = varsOfInterest,
                     selected = "Sample_Type"),
         # select taxonomic rank
         selectInput(inputId = "communityTaxRank", label = "Select Rank:",
                     choices = c("Kingdom", "Phylum", "Class",
                                 "Order", "Family", "Genus",
                                 "Species"),
                     selected = "Phylum"),
         # select abundance filter
         sliderInput(inputId = "abdFilter", label = "Abundance Filter",
                     min = 0, max = 1, step = 0.05, value = 0),
         # update community composition plots
         actionButton(inputId = "updateCommunity",
                      label = "Update Community Plot")
         )
)

#----- React -----#

VariableColumn <- reactive({
  
  variableCol <- input$communityComparison
  
  if (is.null(variableCol)) {
    variableCol <- "Sample_Type"
  }
  
  return(variableCol)
})

# RDP abundance DF
CalculateAbundanceRDP <- eventReactive(input$updateCommunity, {
  
  withProgress(message = "Calculating Abundance...", value = 1, {
    
    abdDF <- MakeAbundanceDF(physeq = physeqListRare$rdp,
                             taxRank = input$communityTaxRank,
                             abundanceFilter = input$abdFilter,
                             pruneMissing = FALSE)
    
    if (dim(abdDF)[1] == 0) {
      stop("No taxa meet this filtering criteria. Trying lowering the Abundance Filter option.", 
           call. = FALSE)
    } else {
      if (input$communityTaxRank == "Species") {
        abdDF <- abdDF %>% 
          mutate("GenusSpecies" = paste(Genus, Species, sep = " "))
      }
    }
  })
  
  if (VariableColumn() == "bad_household") {
    abdDF <- abdDF %>%
      filter(!is.na(bad_household)) %>%
      mutate(bad_household = ifelse(bad_household == TRUE,
                                    yes = "Bad Household",
                                    no = "Good Household"))
  }

  if (VariableColumn() == "Description") {
    abdDF <- abdDF %>%
      filter(Description %in% c("Chronic", "Acute"))
  }

  return(abdDF)
  
}, ignoreNULL = FALSE)


# SILVA abundance DF
CalculateAbundanceSILVA <- eventReactive(input$updateCommunity, {
  
  withProgress(message = "Calculating Abundance...", value = 1, {
    
    abdDF <- MakeAbundanceDF(physeq = physeqListRare$silva,
                             taxRank = input$communityTaxRank,
                             abundanceFilter = input$abdFilter,
                             pruneMissing = FALSE)
    
    if (dim(abdDF)[1] == 0) {
      stop("No taxa meet this filtering criteria. Trying lowering the Abundance Filter option.", 
           call. = FALSE)
    } else {
      if (input$communityTaxRank == "Species") {
        abdDF <- abdDF %>% 
          mutate("GenusSpecies" = paste(Genus, Species, sep = " "))
      }
    }
  })
  
  if (VariableColumn() == "bad_household") {
    abdDF <- abdDF %>% 
      filter(!is.na(bad_household)) %>% 
      mutate(bad_household = ifelse(bad_household == TRUE,
                                    yes = "Bad Household",
                                    no = "Good Household"))
  }
  
  if (VariableColumn() == "Description") {
    abdDF <- abdDF %>% 
      filter(Description %in% c("Chronic", "Acute"))
  }
  
  return(abdDF)
  
}, ignoreNULL = FALSE)
```

```{r standardize-community-plot-colors}
TaxaColors <- eventReactive(input$updateCommunity, {
  
  rdpTaxa <- CalculateAbundanceRDP() %>%
    pull(!!sym(input$communityTaxRank)) %>% 
    unique() %>%
    as.character()
  
  silvaTaxa <- CalculateAbundanceSILVA() %>% 
    pull(!!sym(input$communityTaxRank)) %>% 
    unique() %>% 
    as.character()
  
  allTaxa <- unique(c(rdpTaxa, silvaTaxa))
  
  taxaColors <- hue_pal()(length(allTaxa))
  names(taxaColors) <- allTaxa
  
  return(taxaColors)
    
}, ignoreNULL = FALSE)
```

<h3>RDP</h3>

```{r render-rdp-community}
renderPlotly({
  
  facetBy <- as.formula(paste("~", isolate(input$communityComparison)))
  
  rdpCompPlot <- PlotCommunityComposition(abdDF = CalculateAbundanceRDP(),
                                          taxRank = ifelse(isolate(input$communityTaxRank) == "Species",
                                                           yes = "GenusSpecies",
                                                           no = isolate(input$communityTaxRank)),
                                          facetFormula = facetBy) +
    scale_fill_manual(values = TaxaColors())
  
  ggplotly(rdpCompPlot, height = 500, width = 1000)
})
```

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

<h3>SILVA</h3>

```{r render-silva-community}
renderPlotly({
  
  facetBy <- as.formula(paste("~", isolate(input$communityComparison)))
  
  silvaCompPlot <- PlotCommunityComposition(abdDF = CalculateAbundanceSILVA(),
                                          taxRank = ifelse(isolate(input$communityTaxRank) == "Species",
                                                           yes = "GenusSpecies",
                                                           no = isolate(input$communityTaxRank)),
                                          facetFormula = facetBy) +
    scale_fill_manual(values = TaxaColors())
  
  ggplotly(silvaCompPlot, height = 500, width = 1000)
})
```

<br>

### **Alpha Diversity**

<h3>Alpha Diversity Metrics</h3>

<ul>
  <li>Richness</li>
  <li>Pielou's Evenness</li>
  <li>Shannon Diversity</li>
  <li>Faith's Phylogenetic Diversity (PD)</li>
</ul>

In both the RDP and SILVA data, significant differences in all alpha diversity 
metrics are seen when comparing "bad" households to "good" households. In this
case, the average values for richness, evenness, and diversity are lower on 
average in bad households compared to good.

```{r sampleDataRareAlpha}
alphaIdxList <- list("Richness" = "observed",
                     "Evenness" = "evenness_pielou",
                     "Shannon Diversity" = "diversity_shannon",
                     "Faith's PD" = "PD")

alphaDivList <- map(.x = physeqListRare,
                    .f = ~ microbiome::alpha(x = .x, 
                                             index = c("observed",
                                                       "diversity_shannon",
                                                       "evenness_pielou")))

# add faith's PD
faithsPDList <- map(.x = physeqListRare,
                    .f = ~ pd(samp = as.data.frame(.x@otu_table),
                              tree = .x@phy_tree,
                              include.root = FALSE))

# combine alpha diversity metrics together, then merge with the sample data
allAlphaDivList <- map2(.x = alphaDivList, .y = faithsPDList,
                        .f = ~ merge(.x, .y, by = "row.names", all = TRUE) %>% 
                          column_to_rownames(var = "Row.names"))

# sampleDataRareAlpha contains "observed," "diversity_shannon," 
#   "evenness_pielou," and "PD" columns
sampleDataRareAlpha <- pmap(.l = list(x = sampleDataRare,
                                      y = allAlphaDivList,
                                      by.x = sampleNameColList),
                            .f = base::merge,
                            by.y = "row.names",
                            all = TRUE)
```

```{r make-alpha-plots}
# function to filter alpha diversity data frames for different variables
FilterAlphaDF <- function(df) {
  
  varsNeedFiltering <- c("Household Type" = "bad_household",
                         "CRPS Type" = "Description")
  
  # Household Type
  alphaDFHHType <- df %>% 
    filter(!is.na(bad_household)) %>% 
    mutate(bad_household = ifelse(bad_household == TRUE,
                                  yes = "Bad_Household",
                                  no = "Good_Household"))
  # CRPS Type
  alphaDFCRPSType <- df %>% 
    filter(Description %in% c("Acute", "Chronic"))
  
  alphaDFList <- list("Sample_Type" = df,
                      "bad_household" = alphaDFHHType,
                      "Description" = alphaDFCRPSType)
  
  return(alphaDFList)
}

# list of two elements (RDP and SILVA), each element having three nested lists
alphaDFPlottingList <- map(.x = sampleDataRareAlpha,
                           .f = FilterAlphaDF)
```

```{r alpha-plot-selection}
fluidRow(
  column(5,
         wellPanel(
           selectInput(inputId = "alphaPlot", label = "Select Comparison:",
                       choices = varsOfInterest,
                       selected = "Sample_Type")
           )
         )
  )

AlphaStatMethod <- reactive({
  useMethod <- NULL
  
  if (is.null(input$alphaPlot)) {
    input$alphaPlot <- "Sample_Type"
  }
  
  if (input$alphaPlot == "Sample_Type") {
    useMethod <- "kruskal.test"
  } else {
    useMethod <- "wilcox.test"
  }
  return(useMethod)
})

AlphaPlotsRDP <- reactive({
  alphaPlotsRDP <- pmap(.l = list(yVar = alphaIdxList,
                                  yLab = names(alphaIdxList),
                                  alphaPlotTitle = names(alphaIdxList)),
                        .f = PlotAlphaDiversity,
                        df = alphaDFPlottingList$rdp[[input$alphaPlot]],
                        xVar = input$alphaPlot,
                        statMethod = AlphaStatMethod())
})

AlphaPlotsSILVA <- reactive({
  alphaPlotsSILVA <- pmap(.l = list(yVar = alphaIdxList,
                                  yLab = names(alphaIdxList),
                                  alphaPlotTitle = names(alphaIdxList)),
                        .f = PlotAlphaDiversity,
                        df = alphaDFPlottingList$silva[[input$alphaPlot]],
                        xVar = input$alphaPlot,
                        statMethod = AlphaStatMethod())
})
```

---

<h4>RDP Data:</h4>

<br>

```{r render-alpha-rdp}
renderPlot({
  annotate_figure(grid.arrange(AlphaPlotsRDP()[["Richness"]],
                             AlphaPlotsRDP()[["Faith's PD"]],
                             AlphaPlotsRDP()[["Shannon Diversity"]],
                             ncol = 2, nrow = 2),
                top = text_grob(paste("Alpha Diversity RDP"),
                                size = 14))
}, width = 900, height = 900)
```

<br>
<br>

---

<h4>SILVA Data:</h4>

<br>

```{r render-alpha-silva}
renderPlot({
  annotate_figure(grid.arrange(AlphaPlotsSILVA()[["Richness"]],
                             AlphaPlotsSILVA()[["Faith's PD"]],
                             AlphaPlotsSILVA()[["Shannon Diversity"]],
                             ncol = 2, nrow = 2),
                top = text_grob(paste("Alpha Diversity SILVA"),
                                size = 14))
}, width = 900, height = 900)
```

<br>
<br>

### **Beta Diversity**

<h3>PCoA Plots</h3>

In the RDP and SILVA data, grouping of samples is significant by household status 
(good vs. bad) for both UniFrac and weighted UniFrac distance measures. It is
also significant for CRPS Pain vs. CRPS No Pain for UniFrac distance.

The grouping of samples is also significant between CRPS and BioBank controls
in both the RDP and SILVA data sets when using the weighted UniFrac distance measure.

<br>
<br>

```{r beta-diversity-ui}
fluidRow(
  column(5,
         wellPanel(
           # select groups to compare
           selectInput(inputId = "betaComparison",
                       label = "Select Comparison:",
                       choices = c("CRPS vs. BioBank",
                                   "CRPS vs. Household Controls",
                                   "Acute CRPS vs. Chronic CRPS",
                                   "Good Households vs. Bad Households",
                                   "HHC Pain vs. HHC No Pain",
                                   "CRPS Pain vs. CRPS No Pain"),
                       selected = "CRPS vs. BioBank"),
           # select distance measure
           selectInput(inputId = "betaDistance",
                       label = "Distance Measure:",
                       choices = c("UniFrac" = "unifrac",
                                   "Weighted UniFrac" = "wunifrac"),
                       selected = "UniFrac"),
           # show ellipses?
           checkboxInput(inputId = "betaEllipsesBool",
                         label = "Show Confidence Ellipses",
                         value = FALSE),
           # update
           actionButton(inputId = "betaUpdate",
                        label = "Update Beta Diversity Plots",
                        icon = icon("fas fa-sync-alt"))
           )
         )
  )
```

```{r set-reactive-values-needed-for-beta-div-and-biomarkers}
# Reactive values calculated ahead of time needed for beta diversity plotting

#----- Subsetted Data Frames -----#
# Will use one of the sample data frames because samples are the same in both
#   rdp and silva
df <- sampleDataRare$rdp

# possible subsets:
crpsBB <- df %>% filter(Sample_Type %in% c("CRPS", "BioBank_Control"))
crpsHHC <- df %>% filter(Sample_Type %in% c("CRPS", "HHC"))
crpsType <- df %>% filter(Description %in% c("Acute", "Chronic"))
hhcType <- df %>% filter(!is.na(bad_household_chr))
hhc <- df %>% filter(status_household_description %in% c("HHC_bad_household", "HHC_good_household"))
crps <- df %>% filter(status_household_description %in% c("CRPS_bad_household", "CRPS_good_household"))

# store in reactive value so accessible to shiny:
dfSubsetList <- reactiveValues()
dfSubsetList$`CRPS vs. BioBank` <- crpsBB
dfSubsetList$`CRPS vs. Household Controls` <- crpsHHC
dfSubsetList$`Acute CRPS vs. Chronic CRPS` <- crpsType
dfSubsetList$`Good Households vs. Bad Households` <- hhcType
dfSubsetList$`HHC Pain vs. HHC No Pain` <- hhc
dfSubsetList$`CRPS Pain vs. CRPS No Pain` <- crps

#----- Column Names Corresponding to Selected Comparison -----#
# store the corresponding variable colname in reactive value:
colNameLUT <- reactiveValues()
colNameLUT$`CRPS vs. BioBank` <- "Sample_Type"
colNameLUT$`CRPS vs. Household Controls` <- "Sample_Type"
colNameLUT$`Acute CRPS vs. Chronic CRPS` <- "Description"
colNameLUT$`Good Households vs. Bad Households` <- "bad_household_chr"
colNameLUT$`HHC Pain vs. HHC No Pain` <- "status_household_description"
colNameLUT$`CRPS Pain vs. CRPS No Pain` <- "status_household_description"

#----- Color and Shape Customizations for Plots -----#
colorVector <- reactiveValues()
colorVector$`CRPS vs. BioBank` <- c("CRPS" = "#FB0101", 
                                    "BioBank_Control" = "#FECC66")
colorVector$`CRPS vs. Household Controls` <- c("CRPS" = "#FB0101", 
                                               "HHC" = "#3333CC")
colorVector$`Acute CRPS vs. Chronic CRPS` <- c("Acute" = "#CC66FF", 
                                               "Chronic" = "#408002")
colorVector$`Good Households vs. Bad Households` <- c("Good Household" = "#EB3BEB",
                                                      "Bad Household" = "#FF0700")
colorVector$`HHC Pain vs. HHC No Pain` <- c("HHC_bad_household" = "#FF0700",
                                            "HHC_good_household" = "#EB3BEB")
colorVector$`CRPS Pain vs. CRPS No Pain` <- c("CRPS_bad_household" = "#FF0700",
                                              "CRPS_good_household" = "#EB3BEB")

shapeVector <- reactiveValues()
shapeVector$`CRPS vs. BioBank` <- c("CRPS" = 15, "BioBank_Control" = 16)
shapeVector$`CRPS vs. Household Controls` <- c("CRPS" = 15, "HHC" = 16)
shapeVector$`Acute CRPS vs. Chronic CRPS` <- c("Acute" = 17, "Chronic" = 18)
shapeVector$`Good Households vs. Bad Households` <- c("Good Household" = 18, 
                                                      "Bad Household" = 17)
shapeVector$`HHC Pain vs. HHC No Pain` <- c("HHC_bad_household" = 17,
                                            "HHC_good_household" = 18)
shapeVector$`CRPS Pain vs. CRPS No Pain` <- c("CRPS_bad_household" = 17,
                                              "CRPS_good_household" = 18)
```

```{r beta-diversity-reactive}

#----- Subset Physeq -----#

# Subset both physeq objects in physeqListRare to contain just the samples
#   of interest (returned by input$betaComparison) for plotting.

# For subsetting ps objs:
#   RDP samples in "sample" col
#   SILVA samples in "sequencing_file_name" col

# Set initial value of betaComparison
BetaComparison <- reactive({
  betaComparison <- input$betaComparison
  
  if (is.null(betaComparison)) {
    betaComparison <- "CRPS vs. BioBank"
  }
  
  return(betaComparison)
})

#observe(print(BetaComparison()))

SubsetPhyseq <- eventReactive(input$betaUpdate, {
  
  selectedComparison <- BetaComparison()
  
  # select appropriate samples based on selectedComparison
  # this only needs to be done with one of the two sample data frames since
  #   the set of samples is the same
  
  keepSamplesRDP <- dfSubsetList[[BetaComparison()]] %>% 
    pull(sample) %>% 
    as.character()
  
  keepSamplesSILVA <- dfSubsetList[[BetaComparison()]] %>%
    pull(sequencing_file_name) %>%
    as.character()

  physeqSubsetList <- map2(.x = physeqListRare,
                           .y = list("rdp" = keepSamplesRDP,
                                     "silva" = keepSamplesSILVA),
                           .f = ~ prune_samples(samples = .y,
                                                x = .x))
  
  return(physeqSubsetList)
  
  
}, ignoreNULL = FALSE)

#observe(print(SubsetPhyseq()))

#----- Calculate Ordinations -----#

# Calculate weighted and unweighted ordinations on the physeq objects
OrdinationList <- eventReactive(input$betaUpdate, {
  
  withProgress(message = "Calculating ordinations...", value = 1, {
    set.seed(16738558)
    
    ordList <- map(.x = SubsetPhyseq(),
                   .f = ~ ordinate(physeq = .x,
                                   method = "PCoA",
                                   distance = input$betaDistance))
    })
  
}, ignoreNULL = FALSE)

#----- Calculate ADONIS -----#

AdonisPVals <- eventReactive(input$betaUpdate, {
  
  selectedDistance <- input$betaDistance
  
  if (is.null(input$betaDistance)) {
    selectedDistance <- "unifrac"
  }
  
  withProgress(message = "Calculating ADONIS...", value = 1, {
    
    set.seed(16738558)
    
    adonisResList <- map(.x = SubsetPhyseq(),
                         .f = ~ RunAdonis(physeqObj = .x,
                                          category = colNameLUT[[BetaComparison()]],
                                          distance = selectedDistance))
    
    adonisPVals <- map(.x = adonisResList,
                       .f = ~ .x$aov.tab$`Pr(>F)`[1])
    })

}, ignoreNULL = FALSE)

#observe(print(AdonisPVals()))

```

<h4>RDP Data:</h4>

<br>

```{r render-rdp-beta-plot}
renderPlot({
  
  # for plot title
  distanceString <- NULL
  
  if (isolate(input$betaDistance) == "unifrac") {
    distanceString <- "UniFrac Distance"
  } else {
    distanceString <- "WUniFrac Distance"
  }
  
  # for combining legends
  legendName <- NULL
  
  if (isolate(BetaComparison()) == "CRPS vs. BioBank" | 
      isolate(BetaComparison()) == "CRPS vs. Household Controls") {
    legendName <- "Sample Type"
  } else if (isolate(BetaComparison()) == "Acute CRPS vs. Chronic CRPS") {
    legendName <- "CRPS Type"
  } else if (isolate(BetaComparison()) == "Good Households vs. Bad Households") {
    legendName <- "Household Type"
  } else if (isolate(BetaComparison()) == "HHC Pain vs. HHC No Pain") {
    legendName <- "Household Control Type"
  }
  
  basePlot <- plot_ordination(physeq = SubsetPhyseq()$rdp,
                              ordination = OrdinationList()$rdp,
                              color = colNameLUT[[isolate(BetaComparison())]],
                              shape = colNameLUT[[isolate(BetaComparison())]]) +
    scale_color_manual(values = colorVector[[isolate(BetaComparison())]],
                       name = legendName) +
    scale_shape_manual(values = shapeVector[[isolate(BetaComparison())]],
                       name = legendName) +
    labs(color = isolate(BetaComparison())) +
    theme_bw() +
    geom_point(size = 5) +
    ggtitle(paste0("RDP: ", isolate(BetaComparison()), " (", distanceString, ")"), 
            subtitle = paste0("ADONIS p-val: ", isolate(AdonisPVals())$rdp)) +
    theme(plot.title = element_text(hjust = 0.5, size = 14),
          plot.subtitle = element_text(hjust = 0.5, size = 12),
          axis.text = element_text(size = 12),
          axis.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14))
  
  if (isolate(input$betaEllipsesBool)) {
    basePlot <- basePlot + stat_ellipse(type = "norm")
  }
  
  basePlot
  
}, height = 600, width = 900)
```

<br>
<br>

---

<br>
<br>

<h4>SILVA Data:</h4>

<br>

```{r render-silva-beta-plot}
renderPlot({
  
  # for plot title
  distanceString <- NULL
  
  if (isolate(input$betaDistance) == "unifrac") {
    distanceString <- "UniFrac Distance"
  } else {
    distanceString <- "WUniFrac Distance"
  }
  
  # for combining legends
  legendName <- NULL
  
  if (isolate(BetaComparison()) == "CRPS vs. BioBank" | 
      isolate(BetaComparison()) == "CRPS vs. Household Controls") {
    legendName <- "Sample Type"
  } else if (isolate(BetaComparison()) == "Acute CRPS vs. Chronic CRPS") {
    legendName <- "CRPS Type"
  } else if (isolate(BetaComparison()) == "Good Households vs. Bad Households") {
    legendName <- "Household Type"
  } else if (isolate(BetaComparison()) == "HHC Pain vs. HHC No Pain") {
    legendName <- "Household Control Type"
  }
  
  basePlot <- plot_ordination(physeq = SubsetPhyseq()$silva,
                              ordination = OrdinationList()$silva,
                              color = colNameLUT[[isolate(BetaComparison())]],
                              shape = colNameLUT[[isolate(BetaComparison())]]) +
    scale_color_manual(values = colorVector[[isolate(BetaComparison())]],
                       name = legendName) +
    scale_shape_manual(values = shapeVector[[isolate(BetaComparison())]],
                       name = legendName) +
    labs(color = isolate(BetaComparison())) +
    theme_bw() +
    geom_point(size = 5) +
    ggtitle(paste0("SILVA: ", isolate(BetaComparison()), " (", distanceString, ")"), 
            subtitle = paste0("ADONIS p-val: ", isolate(AdonisPVals())$rdp)) +
    theme(plot.title = element_text(hjust = 0.5, size = 14),
          plot.subtitle = element_text(hjust = 0.5, size = 12),
          axis.text = element_text(size = 12),
          axis.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14))
  
  if (isolate(input$betaEllipsesBool)) {
    basePlot <- basePlot + stat_ellipse(type = "norm")
  }
  
  basePlot
  
}, height = 600, width = 900)
```

<br>
<br>

---

<br>
<br>

### **Pairwise Distances**

<h3>Pairwise UniFrac & Weighted UniFrac Distances of Good and Bad Households</h3>

These plots are comparing the pairwise distances between matched samples (household control + household CRPS sample) from good and bad households. Neither the RDP nor SILVA data set shows a significant difference in the average distance between matched samples in houses with or without pain.

```{r pairwise-distances-rdp, eval=FALSE, include=FALSE}
# RDP Data
set.seed(16738558)

# Calculate the distance objects
rdpDistances <- map(.x = list("UniFrac" = FALSE,
                              "Weighted UniFrac" = TRUE),
                    .f = ~ as.matrix(phyloseq::UniFrac(physeq = physeqListRare$rdp,
                                                       weighted = .x)))
# Melt and remove values of 0 to eliminate samples distances to themselves
rdpDistMelt <- map2(.x = rdpDistances,
                    .y = list("UniFrac" = "UniFrac_Distance",
                              "Weighted UniFrac" = "WUniFrac_Distance"),
                    .f = ~ reshape2::melt(.x, value.name = .y) %>% 
                      mutate(is_self = ifelse(Var1 == Var2,
                                              yes = TRUE,
                                              no = FALSE)) %>% 
                      filter(!is_self) %>% 
                      select(-is_self) %>% 
                      mutate(across(where(is.factor), as.character)) %>% 
                      distinct())

# Remove duplicated measures
rdpDistNoDupes <- map(.x = rdpDistMelt,
                      .f = ~ data.frame(t(base::apply(.x, 1, sort))) %>% 
                        distinct())
names(rdpDistNoDupes$UniFrac) <- c("UniFrac_Distance", "Var1", "Var2")
names(rdpDistNoDupes$`Weighted UniFrac`) <- c("WUniFrac_Distance",
                                              "Var1", "Var2")
rdpDistNoDupes <- map(.x = rdpDistNoDupes,
                      .f = ~ .x %>% 
                        mutate(across(ends_with("Distance"),
                                      as.numeric)))

# remove dupes somehow
# test <- data.frame(V1 = c("A", "A", "B", "B", "C", "C"),
#            V2 = c("B", "C", "A", "C", "A", "B"),
#            n = c(1, 3, 1, 2, 3, 2))
# test2 <- base::apply(test, 1, sort)
# test3 <- data.frame(t(test2))
# test4 <- distinct(test3)

# Add household type metadata to the melted data
rdpHHTypeLUT <- sampleDataRare$rdp %>% 
  select(sample, bad_household_chr) %>% 
  deframe()

rdpDistMeltHH <- map(.x = rdpDistNoDupes,
                     .f = ~ .x %>%
                       mutate(Var1_Type = rdpHHTypeLUT[Var1],
                              Var2_Type = rdpHHTypeLUT[Var2]))

# Select bad-bad HH distances, and good-good HH distances
rdpBadHHDist <- map(.x = rdpDistMeltHH,
                    .f = ~ filter(.x,
                                  Var1_Type == "Bad Household" & 
                                    Var2_Type == "Bad Household"))

rdpGoodHHDist <- map(.x = rdpDistMeltHH,
                     .f = ~ filter(.x,
                                   Var1_Type == "Good Household" &
                                     Var2_Type == "Good Household"))





# rbind good-good and bad-bad for UniFrac & Weighted UniFrac distances
rdpUniFracDist <- rbind(rdpBadHHDist$UniFrac, rdpGoodHHDist$UniFrac)
rdpWUniFracDist <- rbind(rdpBadHHDist$`Weighted UniFrac`,
                         rdpGoodHHDist$`Weighted UniFrac`)

# Make box plot of good versus bad HH average Unifrac & Weighted UniFrac distances
rdpDistBox <- pmap(.l = list(df = list("UniFrac" = rdpUniFracDist,
                                       "Weighted UniFrac" = rdpWUniFracDist),
                             yVar = list("UniFrac" = "UniFrac_Distance",
                                         "Weighted UniFrac" = "WUniFrac_Distance"),
                             plotTitle = names(rdpDistances),
                             label_y = list("UniFrac" = "UniFrac Distance",
                                            "Weighted UniFrac" = "Weighted UniFrac Distance")),
                   .f = PrettyBox,
                   xVar = "Var1_Type",
                   statMethod = "kruskal.test",
                   colorVals = c("Bad Household" = "#FF0700",
                                 "Good Household" = "#EB3BEB"))
# PrettyBox(df = rdpUniFracDist,
#           xVar = "Var1_Type",
#           yVar = "UniFrac_Distance",
#           plotTitle = "UniFrac",
#           colorVals = c("Bad Household" = "#FF0700", "Good Household" = "#EB3BEB"),
#           statMethod = "kruskal.test")



rdpDistBoxArr <- annotate_figure(grid.arrange(rdpDistBox$UniFrac,
                                              rdpDistBox$`Weighted UniFrac`,
                                              ncol = 2, nrow = 1),
                                 top = text_grob("RDP Data\nAvg. Distances of Good and Bad Households", 
                                                 size = 14),
                                 bottom = text_grob("Household Type", size = 12))
rdpDistBoxArr
```

```{r pairwise-distance-plots}
PlotPairwiseDist <- function(physeqObj, sampleDataDF, sampleNameCol,
                             dataName) {
  
  # Calculate the distance objects
  set.seed(16738558)
  
  distances <- map(.x = list("UniFrac" = FALSE, "Weighted UniFrac" = TRUE),
                   .f = ~ as.matrix(phyloseq::UniFrac(physeq = physeqObj,
                                                      weighted = .x)))
  
  # Melt and remove values of 0 to eliminate samples distances to themselves
  distMelt <- map2(.x = distances,
                   .y = list("UniFrac" = "UniFrac_Distance",
                             "Weighted UniFrac" = "WUniFrac_Distance"),
                   .f = ~ reshape2::melt(.x, value.name = .y) %>%
                     mutate(is_self = ifelse(Var1 == Var2,
                                             yes = TRUE, no = FALSE)) %>%
                     filter(!is_self) %>%
                     select(-is_self) %>%
                     mutate(across(where(is.factor), as.character)))
  
  # Remove duplicated measures
  distMeltNoDupes <- map(.x = distMelt,
                         .f = ~ data.frame(t(base::apply(.x, 1, sort))) %>%
                           distinct())
  
  names(distMeltNoDupes$UniFrac) <- c("UniFrac_Distance", "Var1", "Var2")
  names(distMeltNoDupes$`Weighted UniFrac`) <- c("WUniFrac_Distance",
                                                 "Var1", "Var2")
  
  distMeltNoDupes <- map(.x = distMeltNoDupes,
                         .f = ~ .x %>% 
                           mutate(across(ends_with("Distance"), as.numeric)))
  
  # Add household type metadata to the melted data
  hhTypeLUT <- sampleDataDF %>%
    select(!!sym(sampleNameCol), bad_household_chr) %>%
    deframe()
  
  distMeltHH <- map(.x = distMeltNoDupes,
                    .f = ~ .x %>%
                      mutate(Var1_Type = hhTypeLUT[Var1],
                             Var2_Type = hhTypeLUT[Var2]))
  
  # Add household designation so only pairwise members from the same
  #   households are selected
  householdLUT <- sampleDataDF %>% 
    select(!!sym(sampleNameCol), household) %>% 
    deframe()
  
  distMeltHH <- map(.x = distMeltHH,
                    .f = ~ .x %>% 
                      mutate(Var1_HH = householdLUT[Var1],
                             Var2_HH = householdLUT[Var2]))
  
  # Select bad-bad HH distances, and good-good HH distances
  badHHDist <- map(.x = distMeltHH,
                   .f = ~ filter(.x,
                                 Var1_Type == "Bad Household" &
                                   Var2_Type == "Bad Household" &
                                   Var1_HH == Var2_HH))
  
  goodHHDist <- map(.x = distMeltHH,
                    .f = ~ filter(.x,
                                  Var1_Type == "Good Household" &
                                    Var2_Type == "Good Household" &
                                    Var1_HH == Var2_HH))
  
  # rbind good-good and bad-bad for UniFrac & Weighted UniFrac distances
  uniFracDist <- rbind(badHHDist$UniFrac, goodHHDist$UniFrac)
  wUniFracDist <- rbind(badHHDist$`Weighted UniFrac`,
                        goodHHDist$`Weighted UniFrac`)

  
  # Make box plot of good versus bad HH average Unifrac & Weighted UniFrac distances
  distBox <- pmap(.l = list(df = list("UniFrac" = uniFracDist,
                                      "Weighted UniFrac" = wUniFracDist),
                            yVar = list("UniFrac" = "UniFrac_Distance",
                                        "Weighted UniFrac" = "WUniFrac_Distance"),
                            plotTitle = names(distances),
                            label_y = list("UniFrac" = "UniFrac Distance",
                                           "Weighted UniFrac" = "Weighted UniFrac Distance")),
                  .f = PrettyBox,
                  xVar = "Var1_Type",
                  statMethod = "kruskal.test",
                  colorVals = c("Bad Household" = "#FF0700",
                                "Good Household" = "#EB3BEB"))
  
  distBoxArr <- annotate_figure(grid.arrange(distBox$UniFrac,
                                             distBox$`Weighted UniFrac`,
                                             ncol = 2, nrow = 1),
                                top = text_grob(paste(dataName, "Data\nAvg. Distances of Good and Bad Households"),
                                                      size = 14),
                                bottom = text_grob("Household Type", size = 12))
  
  return(distBoxArr)
  
}
```

```{r pairwiseDistPlots, fig.show='hide'}

pairwiseDistPlots <- pmap(.l = list(physeqObj = physeqListRare,
                                    sampleDataDF = sampleDataRare,
                                    sampleNameCol = list("rdp" = "sample",
                                                         "silva" = "sequencing_file_name"),
                                    dataName = list("rdp" = "RDP",
                                                    "silva" = "SILVA")),
                          .f = PlotPairwiseDist)
```

```{r display-pairwise-dist-plots}
pairwiseDistPlots
```

<br>
<br>

### **Biomarker Analysis with DESeq2**

```{r set-reactive-vars-for-biomarker-analysis}
# numerator and denominator options for each comparison
numeratorLUT <- reactiveValues()
numeratorLUT$`CRPS vs. BioBank` <- "CRPS"
numeratorLUT$`CRPS vs. Household Controls` <- "CRPS"
numeratorLUT$`Acute CRPS vs. Chronic CRPS` <- "Acute"
numeratorLUT$`Good Households vs. Bad Households` <- "Bad Household"
numeratorLUT$`HHC Pain vs. HHC No Pain` <- "HHC_bad_household"
numeratorLUT$`CRPS Pain vs. CRPS No Pain` <- "CRPS_bad_household"


denominatorLUT <- reactiveValues()
denominatorLUT$`CRPS vs. BioBank` <- "BioBank_Control"
denominatorLUT$`CRPS vs. Household Controls` <- "HHC"
denominatorLUT$`Acute CRPS vs. Chronic CRPS` <- "Chronic"
denominatorLUT$`Good Households vs. Bad Households` <- "Good Household"
denominatorLUT$`HHC Pain vs. HHC No Pain` <- "HHC_good_household"
denominatorLUT$`CRPS Pain vs. CRPS No Pain` <- "CRPS_good_household"
```

```{r GenerateDESeqResultsModified}
GenerateDESeqResultsModified <- function(physeq, variable, numerator, denominator) {
  
  # Returns DESeq Results as Formal Class "DESeqResults"
  # Create formula from string variable
  formula <- as.formula(paste("~", variable, sep = " "))
  # Convert to deseq data set object
  ds <- phyloseq_to_deseq2(physeq, design = formula)
  # work-around
  counts <- counts(ds)
  geoMeans <- apply(counts, 1,
                      function(row) {if (all(row == 0)) 0 else exp(mean(log(row[row != 0])))})
  dds <- estimateSizeFactors(ds, geoMeans = geoMeans)
  # Run analysis
  ddsAnalysis <- DESeq(dds, test = "Wald", fitType = "local", betaPrior = FALSE)
  # Extract and format results
  ddsResults <- results(ddsAnalysis,
                        contrast = c(variable, numerator, denominator)) 
}
```

```{r deseq-analysis}

#----- UI -----#

fluidRow(
  column(5,
         wellPanel(
           selectInput(inputId = "biomarkerComparison",
                       label = "Select Comparison",
                       choices = c("CRPS vs. BioBank",
                                   "CRPS vs. Household Controls",
                                   "Acute CRPS vs. Chronic CRPS",
                                   "Good Households vs. Bad Households",
                                   "HHC Pain vs. HHC No Pain",
                                   "CRPS Pain vs. CRPS No Pain"),
                       selected = "CRPS vs. BioBank"),
           actionButton(inputId = "updateBiomarkers",
                        label = "Update DESeq Analysis",
                        icon = icon("fas fa-sync-alt"))
           )
         )
  )

#----- React -----#

# Set initial value of biomarkerComparison
BiomarkerComparison <- reactive({
  biomarkerComparison <- input$biomarkerComparison
  
  if (is.null(biomarkerComparison)) {
    biomarkerComparison <- "CRPS vs. BioBank"
  }
  
  return(biomarkerComparison)
})

SubsetPhyseqBiomarkers <- eventReactive(input$updateBiomarkers, {
  
  selectedComparison <- BiomarkerComparison()
  
  # select appropriate samples based on selectedComparison
  # this only needs to be done with one of the two sample data frames since
  #   the set of samples is the same
  
  keepSamplesRDP <- dfSubsetList[[selectedComparison]] %>% 
    pull(sample) %>% 
    as.character()
  
  keepSamplesSILVA <- dfSubsetList[[selectedComparison]] %>%
    pull(sequencing_file_name) %>%
    as.character()

  physeqSubsetList <- map2(.x = physeqListRare,
                           .y = list("rdp" = keepSamplesRDP,
                                     "silva" = keepSamplesSILVA),
                           .f = ~ prune_samples(samples = .y,
                                                x = .x))
  
  return(physeqSubsetList)
  
  
}, ignoreNULL = FALSE)

RunDESeq <- eventReactive(input$updateBiomarkers, {
  
  withProgress(message = "Conducting DESeq2 Analysis...", value = 1, {
    
    colName <- colNameLUT[[BiomarkerComparison()]]
    numeratorVar <- numeratorLUT[[BiomarkerComparison()]]
    denominatorVar <- denominatorLUT[[BiomarkerComparison()]]
    
    deseqResList <- map(.x = SubsetPhyseqBiomarkers(),
                        .f = ~ GenerateDESeqResultsModified(physeq = .x,
                                                           variable = colName,
                                                           numerator = numeratorVar,
                                                           denominator = denominatorVar))
    
  })
  
}, ignoreNULL = FALSE)

DESeqRes <- eventReactive(input$updateBiomarkers, {
  deseqResTables <- map2(.x = SubsetPhyseqBiomarkers(),
                         .y = RunDESeq(),
                         .f = ~ GenerateDESeqResultsTable(physeq = .x,
                                                          sigThreshold = 0.05,
                                                          ddsResults = .y))
}, ignoreNULL = FALSE)
```

<br>

<h4>RDP Data:</h4>

```{r rdp-biomarker-plot}
renderPlotly({
  volcano <- PlotStaticVolcano(physeq = SubsetPhyseqBiomarkers()$rdp,
                               resultsDataTable = DESeqRes()$rdp,
                               sigThreshold = 0.05,
                               plotTitle = paste("Differentially Abundant Taxa\n",
                                                 denominatorLUT[[isolate(BiomarkerComparison())]],
                                                 "(-L2FC) vs",
                                                 numeratorLUT[[isolate(BiomarkerComparison())]],
                                                 "(+ L2FC)\n RDP Data"))
  
  ggplotly(volcano, tooltip = c("Phylum", "Family", "Genus", "Species",
                                "log2FoldChange", "baseMean"),
           height = 550, width = 900)
})
```

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

---

<br>

<h4>SILVA Data</h4>

```{r silva-biomarker-plot}
renderPlotly({
  volcano <- PlotStaticVolcano(physeq = SubsetPhyseqBiomarkers()$silva,
                               resultsDataTable = DESeqRes()$silva,
                               sigThreshold = 0.05,
                               plotTitle = paste("Differentially Abundant Taxa\n",
                                                 denominatorLUT[[isolate(BiomarkerComparison())]],
                                                 "(-L2FC) vs",
                                                 numeratorLUT[[isolate(BiomarkerComparison())]],
                                                 "(+ L2FC)\n SILVA Data"))
  
  ggplotly(volcano, tooltip = c("Phylum", "Family", "Genus", "Species",
                                "log2FoldChange", "baseMean"),
           height = 550, width = 900)
})
```

<br>
<br>

### Session Info

```{r session-info}
Sys.Date()
getwd()
sessionInfo()
```

