---
title: "Human CRPS Combined Cohorts 16S Analysis (Rarefied)"
author: "Rachel Rodgers"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      warning = FALSE,
                      message = FALSE)
```

```{r load-libraries, message=FALSE, include=FALSE}
library("shiny")
library("phyloseq")
library("vegan")
library("gridExtra")
library("knitr")
library("plotly")
library("microbiome")
library("data.table")
library("ggrepel")
library("ggpubr")
library("scales")
library("plyr")
library("DT")
library("picante")
library("gridExtra")
library("tidyverse")

source("./shared_R_scripts/Helper_Functions.R")

options(shiny.sanitize.errors = FALSE)
```

## Overview {.tabset}

---

Analysis of combined human CRPS cohorts:

  * Run 1: 190201 (6 samples)
  * Run 2: 190318 (10 samples)
  * Run 3: 190523 (10 samples)
  * Run 4: 200127 (20 samples)
  * Run 5: 201019 (24 samples)

These cohorts include acute and chronic CRPS patients with their matched household controls, as well as unrelated, healthy "BioBank" controls.

Analyzed:

1. Two groups:
  * CRPS vs HHC (all, acute, chronic)
  * CRPS vs BioBank Controls (all, acute, chronic)

2. Three groups ("sample_type" groups):
  * CRPS
  * HHC
  * BioBank ("healthy")

3. Six groups ("crps_match" groups):
  * Acute CRPS
  * Chronic CRPS
  * Acute HHC
  * Chronic HHC
  * Acute-matched BioBank
  * Chronic-matched BioBank

----

```{r read-in-data, include=FALSE}
physeqRaw <- readRDS("../data/physeqObjects/ps0.rdp_single.RDS")
physeqRaw # 2115 taxa x 69 samples

sampleData <- readRDS("../data/sampleDataModified.RDS")
```

```{r physeqMerged, results='hide', include=FALSE}
# add readsPerSample to sample data prior to merging
readsPerSample <- sample_sums(physeqRaw)
sampleData <- mutate(sampleData, 
                     "readsPerSample" = readsPerSample[sample])

row.names(sampleData) <- sampleData$sample
physeqMerged <- merge_phyloseq(physeqRaw, sample_data(sampleData))
physeqMerged # 2115 x 69
```

### **Quality Control**

#### **Sample Outlier Removal by Unexpected Read Count**

Removing samples having fewer than 1000 reads in the data set, and plotting by group to check for significant differences in average read counts per group.

```{r physeqTrimmed, results='hide'}
lowReadSamples <- sampleData %>% 
  filter(readsPerSample <= 1000) %>% 
  pull(sample) %>% 
  as.character()

keepSamples <- sampleData$sample[!(sampleData$sample %in% lowReadSamples)]

physeqTrimmed <- prune_samples(keepSamples, physeqMerged) %>% 
  RemoveMissingTaxa()

physeqTrimmed # 2108 x 64 samples

sampleDataTrimmed <- as(sample_data(physeqTrimmed), "data.frame")
```
```{r averageReads-physeqTrimmed}
trimmedReadsPerSample <- sample_sums(physeqTrimmed)
summary(trimmedReadsPerSample)
```


```{r sampleGroupsList}
# Creating a list that holds the sample names for each group of comparisons
# Note that the three group (sample_type) and six group (crps_match) comparisons 
#   don't require any subsetting because all of the samples are present in 
#   these groups.

groupNames <- c("sample_type", "crps_match", "household", "acute", "chronic")

groupVars <- c("sample_type", "crps_match", "sample_type", "crps_match", "crps_match")
names(groupVars) <- groupNames

groupVecs <- list(c("BioBank_Control", "CRPS", "HHC"),
                  c("CRPSA", "CRPSC", "HHCA", "HHCC", "NRAM", "NRCM"),
                  c("CRPS", "HHC"),
                  c("NRAM", "CRPSA", "HHCA"),
                  c("NRCM", "CRPSC", "HHCC"))
names(groupVecs) <- groupNames

sampleGroupsDFList <- vector(mode = "list", length = length(groupNames))
sampleGroupsNamesList <- vector(mode = "list", length = length(groupNames))

for (i in 1:length(groupNames)) {

  currentGroupName <- groupNames[i]
  currentGroupVariable <- groupVars[[currentGroupName]]
  currentGroupVector <- groupVecs[[currentGroupName]]
  
  sampleDataSubset <- NULL
  sampleNames <- NULL
  
  if (currentGroupName %in% c("sample_type", "crps_match")) {
    
    sampleDataSubset <- sampleDataTrimmed
    
  } else {
    
    sampleDataSubset <- sampleDataTrimmed %>% 
      filter(!!(sym(currentGroupVariable)) %in% currentGroupVector)
    
  }
  
  sampleGroupsDFList[[i]] <- sampleDataSubset
  names(sampleGroupsDFList)[i] <- currentGroupName
  
  sampleGroupsNamesList[[i]] <- pull(sampleDataSubset, sample)
  names(sampleGroupsNamesList)[i] <- currentGroupName
}

```

```{r fxn-PlotAvgReads}
PlotAvgReads <- function(df, xVar, yVar, groupVar, 
                         plotTitle = NULL, thresh = 1000,
                         labelSamples = FALSE,
                         includeStats = FALSE) {
  
  basePlot <- ggplot(df, aes_string(x = xVar, y = yVar, fill = groupVar)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2, alpha = 0.75) +
    geom_hline(yintercept = thresh, lty = 2) +
    ylab("Reads per Sample") +
    ggtitle(plotTitle) +
    scale_y_log10() +
    theme_pubr() +
    theme(legend.position = "none",
          plot.title = element_text(hjust = 0.5, size = 14),
          axis.title.x = element_blank(),
          axis.text.x = element_text(size = 14))
  
  if (labelSamples == TRUE) {
    # Are there any points below thresh to label?
    lowReadSamples <- subset(df, readsPerSample <= thresh)
    
    if (nrow(lowReadSamples) != 0) {
      
      basePlot <- basePlot + 
      geom_text_repel(aes(label = paste(sample, ": ", readsPerSample)),
                    data = subset(df, readsPerSample <= thresh))
    }
  }
  
  if (includeStats == TRUE) {
    testMethod <- ifelse(n_distinct(df[[xVar]]) > 2,
                         yes = "kruskal.test",
                         no = "wilcox.test")
    
    basePlot <- basePlot +
      stat_compare_means(method = testMethod, label.x.npc = 0)
  }
  
  return(basePlot)

}
```

```{r groupComparisons}
groupComparisons <- c("Broad Conditions (3 groups)" = "sample_type",
                      "Specific Conditions (6 groups)" = "crps_match",
                      "All CRPS vs All HHC" = "household",
                      "All Acute Conditions" = "acute",
                      "All Chronic Conditions" = "chronic")
```

```{r avgReadsPlot, fig.width=11, fig.height=8.5}
AvgReadsPlotList <- pmap(.l = list(df = sampleGroupsDFList,
                                   xVar = groupVars,
                                   groupVar = groupVars,
                                   plotTitle = names(sampleGroupsDFList)),
                         .f = PlotAvgReads,
                         yVar = "readsPerSample",
                         labelSamples = TRUE, includeStats = TRUE)

inputPanel(
  selectInput(inputId = "qcGroup", label = "Groups:",
              choices = groupComparisons,
              selected = "sample_type")
)

renderPlot({
  AvgReadsPlotList[[input$qcGroup]]
})

```

Despite removal of low-read samples, significant difference in average read exists for some sample groupings.  Because a significant difference still exists between groups, we will rarefy the data set.

```{r rarefy, results='hide'}
# Rarefy
physeqRare <- rarefy_even_depth(physeqTrimmed, rngseed = 2236001)
physeqRare # 1,629 taxa
```

```{r save-rarefy, eval=FALSE}
saveRDS(physeqRare, file = "../data/physeqObjects/physeqRare.RDS")
```

Rarefying the dataset drops the total number of taxa from 2,108 to 1,629.

<br>
<br>
<br>

---

#### **Taxa Filtering & Abundance**

Non-bacterial taxa need to be removed from the data set.

```{r filter-non-bacterial, results='hide'}
physeqBacteria <- physeqRare %>% 
  subset_taxa(Kingdom == "Bacteria" & Phylum != "Cyanobacteria/Chloroplast")

# did this deplete any samples?
any(sample_sums(physeqBacteria) == 0) # no

physeqBacteria
```

Removal of non-bacterial taxa further reduces total taxa in the data set from 1,629 to 1,593.

Checking the prevalence and abundance values of different bacterial families to see if there are any low abundance/prevalence taxa.

```{r phylaPrevalence, fig.width=9, fig.height=6}
phylaPrevalence <- TaxRankPrevalence(physeqBacteria, "Phylum")
phylaPrevalence$prevalence_table

phylaPrevalencePlot <- ggplot(phylaPrevalence$prevalence_df,
                              aes(x = TotalAbundance,
                                  y = Prevalence/nsamples(physeqBacteria),
                                  color = Family)) +
  geom_hline(yintercept = 1/207, alpha = 0.5, linetype = 2) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") +
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~ Phylum) +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  ggtitle("Phylum Prevalence in All Samples",
          subtitle = "Colored by Familiy")
phylaPrevalencePlot
```

From the prevalence plots, we can see that several low-abundance/low-prevalence taxa are present in the datset, including the Fusocbacteria, Lentisphaerae, and Syngeristetes.  Downstream analysis (such as biomarker analysis) may benefit from removal of these taxa.

```{r, save-physeqBacteria-for-lefse-input, eval=FALSE, include=FALSE}
# Save the physeqBacteria oject to generate LEfSe input data
saveRDS(physeqBacteria, "../data/physeqObjects/physeqBacteria.RDS")
```

```{r physeqSubsetList}
# Subset physeqBacteria based on the five groups
physeqSubsetList <- map(.x = sampleGroupsNamesList,
                        .f = ~ prune_samples(.x, physeqBacteria) %>% 
                          RemoveMissingTaxa())
```

---

### **16S Analysis**

#### **Community Composition**

```{r community-composition}
#----- UI -----#
inputPanel(
  # select comparison 
  selectInput(inputId = "communityGroup", label = "Groups:",
              choices = groupComparisons, selected = "sample_type"),
  # select taxonomic rank
  selectInput(inputId = "communityRank", label = "Taxonomic Rank:",
              choices = c("Kingdom", "Phylum", "Class", "Order", "Family",
                          "Genus", "Species"),
              selected = "Phylum"),
  # select abundance filter
  sliderInput(inputId = "communityAbdFilt", label = "Abundance Filter:",
              min = 0, max = 1, step = 0.05, value = 0),
  # hide NA?
  checkboxInput(inputId = "communityHideNA", label = "Hide unnamed (NA) taxa",
                value = FALSE)
)

# update community composition plot:
actionButton(inputId = "communityUpdate", 
             label = "Update Community Composition Plot",
             icon = icon("fas fa-sync-alt"))

#----- React -----#
CalculateAbundance <- eventReactive(input$communityUpdate, {
  withProgress(message = "Calculating Abundance...", value = 1, {
                 abdDF <- MakeAbundanceDF(physeq = physeqSubsetList[[input$communityGroup]],
                                          taxRank = input$communityRank,
                                          abundanceFilter = input$communityAbdFilt,
                                          pruneMissing = input$communityHideNA)
                 
                 if (dim(abdDF)[1] == 0) {
                   stop("No taxa meet this filtering criteria. 
                        Try lowering the Abundance Filter option.",
                        call. = FALSE)
                 } else {
                   if (input$communityRank == "Species") {
                     abdDF <- mutate(abdDF,
                                     "GenusSpecies" = paste(Genus, Species, sep = " "),
                                     no = "other")
                   }
                 }
                 abdDF
               })
}, ignoreNULL = FALSE)

GetFacetFormula <- eventReactive(input$communityUpdate,
                                 {facetFormula <- ifelse(input$communityGroup 
                                                         %in% c("sample_type", "household"),
                                                         yes = "~ sample_type",
                                                         no = "~ crps_match")},
                                 ignoreNULL = FALSE)

renderPlotly({
  compPlot <- PlotCommunityComposition(abdDF = CalculateAbundance(),
                                       taxRank = ifelse(isolate(input$communityRank) == "Species",
                                                        yes = "GenusSpecies",
                                                        no = isolate(input$communityRank)),
                                       facetFormula = isolate(GetFacetFormula()))
  
  ggplotly(compPlot, height = 600, width = 1000)
})
```

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

---

#### **Alpha Diversity**

```{r calculate-alpha-idx}
sampleDataBacteria <- as(sample_data(physeqBacteria), "data.frame")

alphaIdxList <- list("Richness" = "observed",
                     "Shannon Diversity" = "diversity_shannon", 
                     "Faith's PD" = "PD")

alphaDiv <- microbiome::alpha(physeqBacteria, index = c("observed",
                                                        "diversity_shannon"))

faithsPD <- pd(as.data.frame(physeqBacteria@otu_table),
               physeqBacteria@phy_tree,
               include.root = FALSE)

allAlphaDiv <- merge(alphaDiv, faithsPD, by = "row.names", all = TRUE)
allAlphaDiv <- column_to_rownames(allAlphaDiv, var = "Row.names")

sampleDataBacteria <- merge(allAlphaDiv, sampleDataBacteria,
                            by = "row.names", all = TRUE) %>% 
  column_to_rownames(var = "Row.names")
```

```{r save-sampleDataBacteria, eval=FALSE, include=FALSE}
write.table(sampleDataBacteria, file = "../analysis/sampleDataBacteria.txt",
            quote = FALSE, sep = "\t", row.names = FALSE)
```


```{r alpha-diversity}
#----- UI -----#
inputPanel(
  # select comparison
  selectInput(inputId = "alphaGroup", label = "Groups:",
              choices = groupComparisons, selected = "sample_type")
)

#----- React -----#
AlphaDF <- reactive({
  dfSubset <- sampleDataBacteria %>% 
    filter(sample %in% sampleGroupsNamesList[[input$alphaGroup]])
  return(dfSubset)
})

MakeAlphaPlots <- reactive({
  
  xVar <- ifelse(input$alphaGroup %in% c("sample_type", "household"),
                 yes = "sample_type", no = "crps_match")
  
  statMethod <- ifelse(n_distinct(AlphaDF()[[xVar]]) > 2,
                       yes = "kruskal.test", no = "wilcox.test")
  
  alphaPlots <- pmap(.l = list(yVar = alphaIdxList, yLab = names(alphaIdxList)),
                     .f = PlotAlphaDiversity,
                     df = AlphaDF(),
                     xVar = xVar,
                     statMethod = statMethod)
  
  alphaPlotsArr <- grid.arrange(alphaPlots$Richness,
                                alphaPlots$`Shannon Diversity`,
                                alphaPlots$`Faith's PD`,
                                ncol = 2, nrow = 2)
  
  return(alphaPlotsArr)
})

renderPlot({MakeAlphaPlots()}, height = 1000, width = 1000)

```

<br>
<br>
<br>

---

#### **Beta Diversity**

```{r beta-diversity}
#----- UI -----#
inputPanel(
  # select comparison
  selectInput(inputId = "betaGroup", label = "Groups:",
              choices = groupComparisons, selected = "sample_type"),
  # select distance method
  radioButtons(inputId = "betaDistance", label = "Distance Method:",
               choices = c("UniFrac" = "unifrac",
                           "Weighted UniFrac" = "wunifrac"),
               selected = "unifrac"),
  # show sample names
  checkboxInput(inputId = "betaShowNames", label = "Show point labels",
                value = FALSE),
  # show confidence ellipses
  checkboxInput(inputId = "betaEllipses", 
                label = "Show confidence ellipses", value = TRUE)
)

#----- React -----#
CalculateOrdination <- reactive({
  withProgress(message = "Calculating ordination...", value = 1, {
    set.seed(13905870)
    ordinationObj <- ordinate(physeqSubsetList[[input$betaGroup]], 
                              method = "PCoA",
                              distance = input$betaDistance)
  })
})

renderPlot({
  
  groupingVar <- ifelse(isolate(input$betaGroup) %in% 
                          c("sample_type", "household"),
                 yes = "sample_type", no = "crps_match")
  
  basePlot <- plot_ordination(physeqSubsetList[[isolate(input$betaGroup)]],
                              ordination = CalculateOrdination(),
                              color = groupingVar, shape = groupingVar) +
    theme_bw() +
    theme(legend.title = element_text(size = 18),
          legend.text = element_text(size = 16),
          axis.title.x = element_text(size = 14),
          axis.title.y = element_text(size = 14),
          axis.text.x = element_text(size = 12),
          axis.text.y = element_text(size = 12)) +
    geom_point(size = 4)
  
  if (input$betaShowNames) {
    basePlot <- basePlot +
      geom_text_repel(aes_string(label = "sample"), color = "grey30", size = 4)
  }
  
  if (input$betaEllipses == TRUE) {
    basePlot <- basePlot + stat_ellipse(type = "norm")
  }
  
  basePlot
  
}, height = 600, width = 900)
```

**ADONIS Results for selected Comparison and Distance Measure:**

```{r beta-adonis-groups}
set.seed(13905870)

renderPrint({
  
  groupingVar <- ifelse(input$betaGroup %in% 
                          c("sample_type", "household"),
                 yes = "sample_type", no = "crps_match")
  
  RunAdonis(physeqObj = physeqSubsetList[[isolate(input$betaGroup)]],
            category = groupingVar, distance = input$betaDistance)
})
```

### **Bacteroides:Firmicutes Ratio (6 groups)**

Calculating the relative abundance of Bacteroides/Firmicutes ratio per sample.

```{r phylumGlomDF, fig.width=11, fig.height=8.5}
phylumGlomDF <- physeqBacteria %>% 
  tax_glom(taxrank = "Phylum", NArm = FALSE) %>% 
  transform_sample_counts(function(x) {x/sum(x)}) %>% 
  psmelt()

# Bacteroides and Firmicutes only
selectPhylaAbd <- phylumGlomDF %>% 
  filter(Phylum %in% c("Bacteroidetes", "Firmicutes")) %>% 
  select(sample, crps_match, Phylum, Abundance)

phylaAbdPivot <- selectPhylaAbd %>% 
  pivot_wider(names_from = Phylum,
              values_from = Abundance) %>% 
  mutate(ratio = Bacteroidetes/Firmicutes)

ratioPlot <- MakeBoxPlot(df = phylaAbdPivot,
                         xVar = "crps_match", yVar = "ratio",
                         label_x = "Group", label_y = "Bacteroides:Firmicutes (rel. abd.)",
                         statMethod = "kruskal.test")
ratioPlot
```

```{r save-phylaAbdPivot, eval=FALSE, include=FALSE}
write.table(phylaAbdPivot, file = "../analysis/phylaAbdPivot.txt",
            quote = FALSE, sep = "\t", row.names = FALSE)
```

```{r save-image, eval=FALSE}
save.image("201129_Human_CRPS_Combined_16S_Analysis.RData")
```


### Session Info

```{r session-info}
Sys.Date()
getwd()
sessionInfo()
```
  